zerver/lib/upload.py

```diff
@@ -399,20 +399,49 @@ def get_public_upload_url(
         # different URL format. Configuring no signature and providing
         # no access key makes `generate_presigned_url` just return the
         # normal public URL for a key.
-        config = Config(signature_version=botocore.UNSIGNED)
-        return self.session.client(
-            "s3",
-            region_name=settings.S3_REGION,
-            endpoint_url=settings.S3_ENDPOINT_URL,
-            config=config,
-        ).generate_presigned_url(
+        #
+        # It unfortunately takes 2ms per query to call
+        # generate_presigned_url, even with our cached boto
+        # client. Since we need to potentially compute hundreds of
+        # avatar URLs in single `GET /messages` request, we instead
+        # back-compute the URL pattern here.
+
+        DUMMY_KEY = "dummy_key_ignored"
+        foo_url = self.get_boto_client().generate_presigned_url(
             ClientMethod="get_object",
             Params={
                 "Bucket": self.avatar_bucket.name,
-                "Key": key,
+                "Key": DUMMY_KEY,
             },
             ExpiresIn=0,
         )
+        parsed = urllib.parse.urlparse(foo_url)
+        base_path = os.path.dirname(parsed.path)
+
+        url_pattern = urllib.parse.urlunparse(
+            parsed._replace(path=os.path.join(base_path, "{key}"))
+        )
+        return url_pattern
```

Comment:
This reinterprets a URL string as a format string that produces a URL string. Any such reinterpretation requires escaping, in case the URL already happened to include format string metacharacters.

Also, this now overlaps the API from `get_public_upload_root_url` (see #18710). I think we should avoid making format strings at all, and just have the API be `urljoin(get_public_upload_root_url(), key)`.

Reply:
Ah, trying handcrafted URL manipulation fails as usual, thanks for catching this! I'll push a PR to use urljoin or something like thaf